### PR TITLE
Spearhead 1944 Compatibility

### PR DESCRIPTION
**Pull request by BrettMayson**
**When merged this pull request will:**
- Add compatibility with SPE
- SPE MGs use primaries, since CSW does not support this, the default behaviour is used (scroll menu actions)
- Requires HEMTT 1.7.2

### CODE REVIEW

File: .hemtt/project.toml [hunk 1]
```diff
@@ -34,5 +34,17 @@ preset = "Hemtt"
...
 [hemtt.launch]
 workshop = [
-    "450814997", # CBA_A3's Workshop ID
+    "450814997", # CBA_A3
+]
+
+[hemtt.launch.spe]
+workshop = [
+    "450814997", # CBA_A3
+    "2369477168", # ADT
```
Comment (BrettMayson) [inline .hemtt/project.toml:43]: forgot to remove, but honestly I think this should stay, and go in the others as well. If the person doesn't have the mod, it just won't load, no problems

Comment (BrettMayson) [reply .hemtt/project.toml:43]: Actually maybe I should add `workshop_optional` or something, and warn people when mods don't exists...

File: optionals/compat_spe/CfgVehicles/turrets.hpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+class SPE_StaticMortar_base: StaticMortar {
+    class Turrets: Turrets {
+        class MainTurret: MainTurret {};
+    };
+    class EventHandlers: EventHandlers {
+        class SPE_StaticWeaponsHandler {
+            init = "";
+        };
+    };
+};
```
Comment (PabstMirror) [inline optionals/compat_spe/CfgVehicles/turrets.hpp:24]:
```
class All {
    class EventHandlers;
};
class AllVehicles: All {};
class Land: AllVehicles {};
class LandVehicle: Land {};
class StaticWeapon: LandVehicle {
    class ACE_Actions {
        class ACE_MainActions;
    };
    class Turrets {
        class MainTurret;
    };
    class UserActions;
};

// --- Mortars -----------------------------------------------------------------
class StaticMortar: StaticWeapon {
    class Turrets: Turrets {
        class MainTurret: MainTurret {};
    };
};
class SPE_StaticMortar_base: StaticMortar {
    class Turrets: Turrets {
        class MainTurret: MainTurret {};
    };
    class EventHandlers: EventHandlers {
        class SPE_StaticWeaponsHandler {
            init = "";
        };
    };
};
```
I think this is what we need, not tested

Comment (BrettMayson) [reply optionals/compat_spe/CfgVehicles/turrets.hpp:24]: Ah yeah I forgot to circle back to that, wanted to check where they were defined

Comment (Godlikehobbit) [reply optionals/compat_spe/CfgVehicles/turrets.hpp:24]: I tested Pabst's changes, worked perfectly

File: optionals/compat_spe/CfgVehicles/wheeled.hpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    EGVAR(refuel,hooks)[] = {{-0.23,-2.58,-0.59}};
+    EGVAR(refuel,fuelCargo) = 2000;
+};
+
+class SPE_OpelBlitz_Ambulance: SPE_OpelBlitz_base {
+    EGVAR(medical,medicClass) = 1;
+};
+
+class SPE_OpelBlitz_Repair: SPE_OpelBlitz_base {
+    transportRepair = 0;
```
Comment (PabstMirror) [inline optionals/compat_spe/CfgVehicles/wheeled.hpp:56]:
```suggestion
```
I think we can drop all these after https://github.com/acemod/ACE3/pull/8070

File: optionals/compat_spe/CfgVehicles/wheeled.hpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    EGVAR(repair,canRepair) = 1;
+    transportAmmo = 0;
```
Comment (PabstMirror) [inline optionals/compat_spe/CfgVehicles/wheeled.hpp:58]:
```suggestion
```

**Review by ZluskeN — CHANGES_REQUESTED** [review]

**Review by veteran29 — CHANGES_REQUESTED** [review]

File: optionals/compat_spe/CfgVehicles/wheeled.hpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    EGVAR(vehicle_damage,engineDetonationProb) = 0.03;
+    EGVAR(vehicle_damage,hullFireProb) = 0.3;
+    EGVAR(vehicle_damage,turretFireProb) = 0.1;
+    EGVAR(vehicle_damage,engineFireProb) = 0.2;
+    EGVAR(vehicle_damage,detonationDuringFireProb) = 0.2;
+    EGVAR(vehicle_damage,canHaveFireRing) = 1;
+};
+
+
+// W H E E L E D - AXIS
```
Comment (veteran29) [inline optionals/compat_spe/CfgVehicles/wheeled.hpp:40]:
```suggestion
// WHEELED - AXIS
```

File: optionals/compat_spe/CfgVehicles/wheeled.hpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    transportAmmo = 0;
+    EGVAR(rearm,defaultSupply) = 1200;
+};
+
+class SPE_OpelBlitz_Ammo: SPE_OpelBlitz_base {
+    transportAmmo = 0;
+    EGVAR(rearm,defaultSupply) = 1200;
+};
+
+// W H E E L E D ALLIED FORCES
```
Comment (veteran29) [inline optionals/compat_spe/CfgVehicles/wheeled.hpp:67]:
```suggestion
// WHEELED - ALLIED FORCES
```

File: optionals/compat_spe/CfgWeapons/csw.hpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
+class Launcher;
+class Launcher_Base_F: Launcher {
+    class WeaponSlotsInfo;
+};
+
+class SPE_Slung_Static_Weapon_Base: Launcher_Base_F {};
```
Comment (veteran29) [inline optionals/compat_spe/CfgWeapons/csw.hpp:6]:
```suggestion
class SPE_Slung_Static_Weapon_Base;
```

File: optionals/compat_spe/CfgWeapons/helmets.hpp [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+#define HEARING_PROTECTION_VICCREW EGVAR(hearing,protection) = 0.85; EGVAR(hearing,lowerVolume) = 0.6;
+#define HEARING_PROTECTION_EARMUFF EGVAR(hearing,protection) = 0.75; EGVAR(hearing,lowerVolume) = 0.5;
+#define HEARING_PROTECTION_PELTOR EGVAR(hearing,protection) = 0.75; EGVAR(hearing,lowerVolume) = 0;
+
+class ItemCore;
+class H_HelmetB: ItemCore {
+    class ItemInfo;
+};
+class H_SPE_HelmetB: H_HelmetB {};
+class H_SPE_Hat: H_SPE_HelmetB {};
```
Comment (veteran29) [inline optionals/compat_spe/CfgWeapons/helmets.hpp:11]:
```suggestion
class H_SPE_HelmetB;
class H_SPE_Hat: H_SPE_HelmetB {};
```

File: optionals/compat_spe/CfgWeapons/launchers.hpp [hunk 9]
```diff
@@ -0,0 +1,21 @@
+class LauncherCore;
+class Launcher: LauncherCore {};
+class Launcher_Base_F: Launcher {};
```
Comment (veteran29) [inline optionals/compat_spe/CfgWeapons/launchers.hpp:3]:
```suggestion
class Launcher_Base_F;
```

File: optionals/compat_spe/CfgWeapons/launchers.hpp [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    EGVAR(overpressure,range) = 10;
+};
+class SPE_PzFaust_30m: SPE_LAUNCHER {
+    EGVAR(overpressure,angle) = 60;
+    EGVAR(overpressure,damage) = 0.7;
+    EGVAR(overpressure,priority) = 1;
+    EGVAR(overpressure,range) = 10;
+};
+class SPE_PzFaust_60m: SPE_PzFaust_30m {};
+class SPE_Faustpatrone: SPE_PzFaust_30m {};
```
Comment (veteran29) [inline optionals/compat_spe/CfgWeapons/launchers.hpp:21]:
```suggestion
```

**Review by veteran29 — CHANGES_REQUESTED** [review]

File: optionals/compat_spe/config.cpp [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+            "ace_rearm",
+            "ace_refuel",
+            "ace_repair",
+            "ww2_spe_assets_c_weapons_infantryweapons_c",
+            "ww2_spe_assets_c_vehicles_staticweapons_c",
+            "ww2_spe_assets_c_vehicles_weapons_c",
+            "ww2_spe_core_f_system_staticweapons_f",
+            "ww2_spe_core_c_core_c_eventhandlers"
+        };
+        author = ECSTRING(common, ACETeam);
```
Comment (veteran29) [inline optionals/compat_spe/config.cpp:22]:
```suggestion
        author = ECSTRING(common,ACETeam);
```

File: optionals/compat_spe/config.cpp [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        authors[] = {
+            "sancron",
+            "nomisum",
+            "coldfront15/Henderson",
+            "BrettMayson"
+        };
+        url = ECSTRING(main, URL);
```
Comment (veteran29) [inline optionals/compat_spe/config.cpp:29]:
```suggestion
        url = ECSTRING(main,URL);
```

File: optionals/compat_spe/script_version.hpp [hunk 13]
```diff
@@ -0,0 +1,4 @@
+#define MAJOR 0
+#define MINOR 5
+#define PATCHLVL 15
+#define BUILD 2
```
Comment (veteran29) [inline optionals/compat_spe/script_version.hpp:4]: Version from main mod should be used.

**Review by veteran29 — CHANGES_REQUESTED** [review]

**Comment by PabstMirror** [comment]
moved csw/explosives/refuel to sub configs
should be good to go, only tested with `@Spearhead 1944 -  Creator DLC Compatibility Data for Non-Owners`

**Review by PabstMirror — APPROVED** [review]

**Review by jonpas — APPROVED** [review]

**Review by ZluskeN — APPROVED** [review]

File: addons/compat_spe/CfgWeapons/weapons.hpp [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+class SPE_Sten_Mk2: SPE_SMG {
+    ACE_barrelTwist = 254;
+    ACE_barrelLength = 196;
+    EGVAR(overheating,closedBolt) = 0;
+};
+class SPE_Sten_Mk2_Suppressed: SPE_Sten_Mk2 {
+    ACE_barrelTwist = 254;
+    ACE_barrelLength = 196;
+    EGVAR(overheating,closedBolt) = 0;
+};
```
Comment (rautamiekka) [inline addons/compat_spe/CfgWeapons/weapons.hpp:137]: Nothing changes in this class, so there's no point in re-defining the values.

File: addons/compat_spe/CfgWeapons/weapons.hpp [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+    EGVAR(overheating,closedBolt) = 0;
+    EGVAR(overheating,slowdownFactor) = 1;
+    EGVAR(overheating,allowSwapBarrel) = 1;
+    EGVAR(overheating,dispersion) = 0.25;
+};
+
+// SELF LOADING RIFLES
+class SPE_RIFLE;
+class SPE_G43: SPE_RIFLE {
+    ACE_Overheating_JamChance = 0.0003;
```
Comment (rautamiekka) [inline addons/compat_spe/CfgWeapons/weapons.hpp:61]:
```suggestion
    EGVAR(overheating,JamChance) = 0.0003;
```

File: addons/compat_spe/CfgWeapons/weapons.hpp [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+    ACE_barrelTwist = 254; // unknown set to 1:10
+    ACE_barrelLength = 550;
+};
+class SPE_K98: SPE_RIFLE {
+    ACE_Overheating_JamChance = 0.0003;
```
Comment (rautamiekka) [inline addons/compat_spe/CfgWeapons/weapons.hpp:66]:
```suggestion
    EGVAR(overheating,JamChance) = 0.0003;
```

File: addons/compat_spe/CfgWeapons/weapons.hpp [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+    ACE_barrelTwist = 254; // unknown set to 1:10
+    ACE_barrelLength = 550;
+};
+class SPE_M1_Carbine: SPE_RIFLE {
+    ACE_Overheating_JamChance = 0.0003;
```
Comment (rautamiekka) [inline addons/compat_spe/CfgWeapons/weapons.hpp:71]:
```suggestion
    EGVAR(overheating,JamChance) = 0.0003;
```

File: addons/compat_spe/CfgWeapons/weapons.hpp [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+    ACE_barrelTwist = 254; // unknown set to 1:10
+    ACE_barrelLength = 460;
+};
+class SPE_M1_Garand: SPE_RIFLE {
+    ACE_Overheating_JamChance = 0.0003;
```
Comment (rautamiekka) [inline addons/compat_spe/CfgWeapons/weapons.hpp:76]:
```suggestion
    EGVAR(overheating,JamChance) = 0.0003;
```

File: addons/compat_spe/CfgWeapons/weapons.hpp [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+    ACE_barrelLength = 609.6;
+};
+class SPE_M1903A3_Springfield: SPE_RIFLE {
+    ACE_barrelTwist = 254;
+    ACE_barrelLength = 610;
+};
+
+// ASSAULT RIFLE
+class SPE_STG44: SPE_RIFLE {
+    ACE_Overheating_JamChance = 0.0015;
```
Comment (rautamiekka) [inline addons/compat_spe/CfgWeapons/weapons.hpp:87]:
```suggestion
    EGVAR(overheating,JamChance) = 0.0015;
```

**Comment by PabstMirror** [comment]
build may fail because updated `.github/workflows/pboproject.yml` won't take effect until after merge

**Review by rautamiekka — CHANGES_REQUESTED** [review]

File: addons/compat_spe/CfgWeapons/helmets.hpp [hunk 20]
```diff
@@ -0,0 +1,147 @@
+#define HEARING_PROTECTION_OPEN EGVAR(hearing,protection) = 0; EGVAR(hearing,lowerVolume) = 0;
+#define HEARING_PROTECTION_VICCREW EGVAR(hearing,protection) = 0.85; EGVAR(hearing,lowerVolume) = 0.6;
+#define HEARING_PROTECTION_EARMUFF EGVAR(hearing,protection) = 0.75; EGVAR(hearing,lowerVolume) = 0.5;
+#define HEARING_PROTECTION_PELTOR EGVAR(hearing,protection) = 0.75; EGVAR(hearing,lowerVolume) = 0;
```
Comment (jonpas) [inline addons/compat_spe/CfgWeapons/helmets.hpp:4]:
```suggestion
#define HEARING_PROTECTION_OPEN EGVAR(hearing,protection) = 0; EGVAR(hearing,lowerVolume) = 0
#define HEARING_PROTECTION_VICCREW EGVAR(hearing,protection) = 0.85; EGVAR(hearing,lowerVolume) = 0.6
#define HEARING_PROTECTION_EARMUFF EGVAR(hearing,protection) = 0.75; EGVAR(hearing,lowerVolume) = 0.5
#define HEARING_PROTECTION_PELTOR EGVAR(hearing,protection) = 0.75; EGVAR(hearing,lowerVolume) = 0
```
Macro use should have a trailing semi-colon instead.

Comment (PabstMirror) [reply addons/compat_spe/CfgWeapons/helmets.hpp]: replaced all usage of these macros in a single define in \hearing

File: addons/compat_gm/script_component.hpp [hunk 21]
```diff
@@ -3,3 +3,6 @@
 
 #include "\z\ace\addons\main\script_mod.hpp"
 #include "\z\ace\addons\main\script_macros.hpp"
+
+#include "\z\ace\addons\csw\script_config_macros_csw.hpp"
+#include "\z\ace\addons\hearing\script_macros_hearingProtection.hpp"
```
Comment (jonpas) [inline addons/compat_gm/script_component.hpp:8]: This causes a requirement on CSW and Hearing.

**Review by jonpas — COMMENTED** [review]
This adds CSW and Hearing includes to base compats instead of sub-components.

File: addons/compat_gm/script_component.hpp (hunk 21 above)
Comment (PabstMirror) [reply addons/compat_gm/script_component.hpp]: moved to config.cpp

**Comment by jonpas** [comment]
```
  Log P:\temp\compat_spe.packing.log tail:
    9262: rapWarning: **********missing file(s)***************
    9263: 
    9264: \z\ace\addons\compat_spe\compat_spe_explosives\config.cpp Rapify:rapWarning: **********missing file(s)***************
```

File: addons/compat_gm/script_component.hpp (hunk 21 above)
Comment (jonpas) [reply addons/compat_gm/script_component.hpp]: How does that fix it? :thinking: 

CSW include should be in a sub-component. But I guess GM doesn't have one? Although I guess those are only needed for building?

**Comment by PabstMirror** [comment]
will build ok on merge (changes to pboproject.yml don't take effect inside the pr)

File: addons/compat_spe/compat_spe_csw/CfgVehicles.hpp [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+                init = "";
+            };
+        };
+    };
+
+    class SPE_US_Mortar_base: SPE_StaticMortar_base {};
+    class SPE_M1_81: SPE_US_Mortar_base {
+        class Turrets: Turrets {
+            class MainTurret: MainTurret {
+                magazines[] = {};
```
Comment (veteran29) [inline addons/compat_spe/compat_spe_csw/CfgVehicles.hpp:39]: Why we're zeroing magazines array here? Won't this make the mortar unusable/empty, CSW compat does not need this IIRC?

**Review by veteran29 — DISMISSED** [review]

File: addons/compat_spe/compat_spe_csw/CfgVehicles.hpp (hunk 22 above)
Comment (LinkIsGrim) [reply addons/compat_spe/compat_spe_csw/CfgVehicles.hpp:39]: Yeah, CSW shouldn't need this

File: addons/compat_gm/script_component.hpp (hunk 21 above)
Comment (LinkIsGrim) [reply addons/compat_gm/script_component.hpp]: > How does that fix it? 🤔

Requirement is only there on building.

File: addons/compat_spe/compat_spe_csw/CfgVehicles.hpp (hunk 22 above)
Comment (PabstMirror) [reply addons/compat_spe/compat_spe_csw/CfgVehicles.hpp:39]: csw was designed to be a runtime setting
so we shouldn't be doing this or fully removing actions (but we might have to modify them to be compat with csw)
```
        class UserActions {
            delete Prepare_Illu_Selected;
```
maybe we just comment out
```
#include "CfgVehicles.hpp"
#include "CfgWeapons.hpp"
```
for now and finish this sub-component up later

File: .hemtt/project.toml [hunk 23]
```diff
@@ -34,7 +34,18 @@ preset = "Hemtt"
...
+    "450814997", # CBA_A3
+]
+
+[hemtt.launch.spe]
+workshop = [
+    "450814997", # CBA_A3
+]
+optionals = [
+    "compat_spe"
+]
```
Comment (PabstMirror) [inline .hemtt/project.toml:46]: pretty sure we don't need this?
```suggestion
```

**Review by PabstMirror — APPROVED** [review]

**Review by jonpas — APPROVED** [review]

**Pull request merged**